sdk/core/azure-core/inc/azure/core/hash.hpp

```diff
@@ -0,0 +1,146 @@
+
+    /**
+     * @brief Computes the hash value of the specified binary input data, including any previously
+     * appended.
+     * @param data The pointer to binary data to compute the hash value for.
+     * @param length The size of the data provided.
+     * @return The computed hash value corresponding to the input provided including any previously
+     * appended.
+     */
+    virtual std::vector<uint8_t> OnFinal(const uint8_t* data, std::size_t length) = 0;
+
+  public:
+    /**
+     * @brief Used to append partial binary input data to compute the hash in a streaming fashion.
+     * @remark Once all the data has been added, call #Final() to get the computed hash value.
+     * @remark Do not call this function after a call to #Final().
+     * @param data The pointer to the current block of binary data that is used for hash
+     * calculation.
+     * @param length The size of the data provided.
+     */
+    virtual void Append(const uint8_t* data, std::size_t length)
+    {
+      if (!data && length != 0)
+      {
+        throw std::invalid_argument("Length cannot be non-zero if the data pointer is null.");
+      }
+      if (m_isdone)
+      {
+        throw std::runtime_error("Cannot call Append after calling Final().");
+      };
+      OnAppend(data, length);
+    }
+
+    /**
+     * @brief Computes the hash value of the specified binary input data, including any previously
+     * appended.
+     * @remark Do not call this function multiple times.
+     * @param data The pointer to the last block of binary data to compute the hash value for.
+     * @param length The size of the data provided.
+     * @return The computed hash value corresponding to the input provided, including any previously
+     * appended.
+     */
+    virtual std::vector<uint8_t> Final(const uint8_t* data, std::size_t length)
+    {
+      if (!data && length != 0)
+      {
+        throw std::invalid_argument("Length cannot be non-zero if the data pointer is null.");
+      }
+      if (m_isdone)
+      {
+        throw std::runtime_error("Cannot call Final() multiple times.");
+      };
+      m_isdone = true;
+      return OnFinal(data, length);
+    }
+
+    /**
+     * @brief Computes the hash value of all the binary input data appended to the instance so far.
+     * @remark Use #Append() to add more partial data before calling this function.
+     * @remark Do not call this function multiple times.
+     * @return The computed hash value corresponding to the input provided.
+     */
+    virtual std::vector<uint8_t> Final() { return Final(nullptr, 0); };
+
+    /**
+     * @brief Cleanup any state when destroying the instance of @Hash.
+     */
+    virtual ~Hash(){};
+
+  private:
+    bool m_isdone = false;
+  };
+
+  /**
+   * @brief Represents the class for the MD5 hash function which maps binary data of an arbitrary
+   * length to small binary data of a fixed length.
+   */
+  class Md5Hash : public Hash {
```

Comment:
Wouldn't it be better to let this class be in a new hpp?  Is is ok to expect all Hash classes to live here?

Reply:
I think it's fine. Similar to bodystream.